### ISSUE TITLE
`emmet.triggerExpansionOnTab` enters visual mode

### ISSUE THREAD

**Issue by cusxio** [VSCodeVim/Vim#5884]
**Describe the bug**

When `emmet.triggerExpansionOnTab` is triggered via `TAB`, the editor enters visual mode.

**To Reproduce**

1. Open a new `.html` file.
2. In insert mode, type `div`.
3. Press `Tab`.
4. The editor enters visual mode.

---

**Expected behavior**

The editor should remain in insert mode.

---

**Screenshots**

![Kapture 2020-12-14 at 16 08 25](https://user-images.githubusercontent.com/6487613/102056349-317f3900-3e27-11eb-9931-df41ef8f0cbf.gif)

---

**Environment (please complete the following information):**

- Extension (VsCodeVim) version: 1.18.5
- VSCode version: 1.52.0
- OS: macOS Catalina

---

**Additional context**

Related to: https://github.com/VSCodeVim/Vim/issues/5474

The PR referenced in the issue above did not fix the issue.

**Comment by berknam** [comment]
I can't reproduce this on Windows 10. Can you show what happens when you do that with Vim disabled? Do you have any remaps in "vim.insertModeKeybindings" for the emmet trigger expansion command?

**Comment by cusxio** [comment]
> Can you show what happens when you do that with Vim disabled?

![Kapture 2020-12-16 at 10 00 39](https://user-images.githubusercontent.com/6487613/102295100-a10b3a80-3f85-11eb-9720-f33059790b98.gif)

>  Do you have any remaps in "vim.insertModeKeybindings" for the emmet trigger expansion command?

No I don't.

The only settings enabled are

```json
{
    "emmet.showExpandedAbbreviation": "never",
    "emmet.triggerExpansionOnTab": true
}
```

I've disabled all extension and only enabled the vim extension and those settings. Note that the editor enters visual mode when `Tab` is pressed.

![Kapture 2020-12-16 at 10 04 48](https://user-images.githubusercontent.com/6487613/102295382-2c84cb80-3f86-11eb-88d8-6c84674bed58.gif)

#### Side Note

I noticed another thing, if you enable `emmet.showExpandedAbbreviation`, then pressing `Tab` works as expected **for known HTML elements**, **but unknown element element, where the suggestion is not shown**, then vim enters visual mode.

**Comment by evolify** [comment]
The same issue

**Comment by berknam** [comment]
@cusxio @evolify Are you using the Insiders version?

**Comment by berknam** [comment]
I am able to reproduce this on the Insiders version. I'm also able to reproduce this on normal version but only if using the commandId `emmet.expandSelection` instead of the one that is bound to TAB by default that is `editor.emmet.action.expandAbbreviation`.

From what I can tell, [this commit](https://github.com/microsoft/vscode/commit/ffacef4727ef6896823b895d5684f7e9e3df27fb#diff-bd48396a1c0e916db7bee11a3b9b514af7b9ff2764a1e4848250dfe87c7f6b6b) is the one that reintroduces the issue on the Insiders version by reverting the previous commit.

There has been an update on the emmet extension which added the two expand commands, the old `emmet.expandSelection` that has the issue and the new `editor.emmet.action.expandAbbreviation` that works fine.

The `emmet.expandSelection` command seems to be selecting the word before removing it and inserting the snippet. That selection change makes Vim go to Visual mode.

@rzhao271 Hey! Sorry to bother. But can you help out here with this?

**Comment by cusxio** [comment]
@berknam 

I personally can reproduce it on both Insiders and non-Insiders version. And am sure that the expansion occurs via `editor.emmet.action.expandAbbreviation`

Btw, per https://github.com/VSCodeVim/Vim/issues/5495, I downgraded to 1.6 and the issue no longer persist.

So I suspect there should be a commit between 1.6 and 1.7 that causes this.

**Comment by berknam** [comment]
> So I suspect there should be a commit between 1.6 and 1.7 that causes this.

There was a change that makes it so that when vscode creates a selection Vim goes to visual mode.

But there has been a change in emmet as well that is creating a selection before expanding the abbreviation and that is what is causing the issue.

For me on windows 10, vscode 1.52.1 and Vim 1.18.5 with emmet settings by default if I type `div` and then press TAB it expands to `<div>|</div>` and keeps me on InsertMode.

If I enable `emmet.triggerExpansionOnTab` the same thing happens. If I set `emmet.showExpandedAbbreviation` to `never` then I can't get the abbreviation to expand at all using TAB even if `emmet.triggerExpansionOnTab` is enabled. The only way I can get it to expand is to either call the `Emmet: Expand Abbreviation` from command palette or by using a keybinding to the `emmet.expandAbbreviation` command or even by using a remap or a keybinding to `editor.emmet.action.expandAbbreviation`. And in all these cases I can see that a selection of the word `div` is being made and only after it will it expand the abbreviation, this is what causes the issue.

For some reason the default TAB keybinding to `editor.emmet.action.expandAbbreviation` doesn't work when `emmet.showExpandedAbbreviation` is set to `never`. But when it is set to always the default TAB keybinding to `editor.emmet.action.expandAbbreviation` works differently from a user keybinding or remap to that same command by not creating that selection first. (this is weird 😕 )

@cusxio Can you please check you "Keyboard Shortcuts" and search for `emmet expand` and post a printscreen of that? Is the default TAB keybinding set to `editor.emmet.action.expandAbbreviation`? Does it show the source as 'Default' or 'User'?

**Comment by cusxio** [comment]
> If I enable emmet.triggerExpansionOnTab the same thing happens. If I set emmet.showExpandedAbbreviation to never then I can't get the abbreviation to expand at all using TAB even if emmet.triggerExpansionOnTab is enabled. The only way I can get it to expand is to either call the Emmet: Expand Abbreviation from command palette or by using a keybinding to the emmet.expandAbbreviation command or even by using a remap or a keybinding to editor.emmet.action.expandAbbreviation. And in all these cases I can see that a selection of the word div is being made and only after it will it expand the abbreviation, this is what causes the issue.

Your debugging result seems consistent with mine.

![Kapture 2020-12-24 at 08 42 44](https://user-images.githubusercontent.com/6487613/103046663-0aadc900-45c4-11eb-8a81-c9edbcde180f.gif)

When `emmet.showExpandedAbbreviation: never`, the "suggestion" box never appears, and because it never appears, there's no way to expand the abbreviation unless you call `Emmet: Expand Abbreviation` (i.e. through the command palette).

---

> @cusxio Can you please check you "Keyboard Shortcuts" and search for emmet expand and post a printscreen of that? Is the default TAB keybinding set to editor.emmet.action.expandAbbreviation? Does it show the source as 'Default' or 'User'?

It's the default.

![Screenshot 2020-12-24 at 8 35 13 AM](https://user-images.githubusercontent.com/6487613/103046822-8c055b80-45c4-11eb-8564-1ee0c0791390.png)

---

> For some reason the default TAB keybinding to editor.emmet.action.expandAbbreviation doesn't work when emmet.showExpandedAbbreviation is set to never. But when it is set to always the default TAB keybinding to editor.emmet.action.expandAbbreviation works differently from a user keybinding or remap to that same command by not creating that selection first. (this is weird 😕 )

I think there's a subtle but important difference here.

1. The default `editor.emmet.action.expandAbbreviation` shortcut requires `emmet.triggerExpansionOnTab: true` (as seen in the image above). Without it, it will not work.

2. `emmet.showExpandedAbbreviation: never` will disable the "suggestion" box.

---

TLDR:

I think there's 2 ways to trigger emmet expand abbreviation.

1. Triggering via the suggestion box, works as intended.

2. Triggering via calling `editor.emmet.action.expandAbbreviation` is broken.

**Comment by rzhao271** [comment]
Hi, the two commands currently work as follows:

- emmet.expandAbbreviation is a command within the Emmet extension
- editor.emmet.action.expandAbbreviation is a command (within the VS Code core for now) that does some preprocessing before calling emmet.expandAbbreviation.

Therefore, editor.emmet.action.expandAbbreviation is the command that should be called by the tab key press. That command should also be the one called by the suggestion box/command palette by default.

**Comment by rzhao271** [comment]
@berknam

Both methods of expansion go into visual mode for me now. 
Is there an easy way to make the extension go back into insert mode, such as by providing a new empty selection after the expansion?

Otherwise, we have the following issues:
- Expand abbreviation selects the abbreviation to replace it with the expanded text
- Wrap with Abbreviation selects the text to then wrap around it with some expanded abbreviation

I'm not sure how these operations should be handled without changing the selected text.

**Comment by notiv-nt** [comment]
@berknam i've found, the error does not apply to emmet

This commit causes the problem https://github.com/VSCodeVim/Vim/commit/35d06e6780fbde4624081e47ddfc95cd60a1451d (Jul 25)

https://user-images.githubusercontent.com/10743009/105640384-c9445e00-5e8e-11eb-9d1f-1a4ff327bff7.mov

1 commit before: https://github.com/VSCodeVim/Vim/commit/94b3c684ec708c3d553bb5e0f2478341df4b743d (Jul 24):

https://user-images.githubusercontent.com/10743009/105640382-c5184080-5e8e-11eb-9352-c6669be3bf8d.mov

**Comment by evolify** [comment]
any progress now ?

**Comment by mrmurphy** [comment]
This is still an issue in 1.21.1 in case anyone was wondering. Thanks for your work! <3

**Comment by devstefancho** [comment]
I still have the same issue, and it's really bothersome.
How do you guys use emmet with this issue, what is the better way to this issue, currently? 
any solution?

**Comment by notiv-nt** [comment]
@devstefancho I switched to amVim

**Comment by rehack** [comment]
How to fix it

**Comment by devstefancho** [comment]
@rehack Cool!, I install another version(1.16.0) and The problem is gone like magic.

This is [How to switch to another version](https://stackoverflow.com/a/66134500/11650728)

<img alt="how to change" src="https://user-images.githubusercontent.com/61320923/130339494-b0ebfa64-bd4a-4242-875e-3997019b3f3f.png" width=300/>

**Comment by orlandoxu** [comment]
This is still an issue in v1.21.6 in case anyone was wondering. Thanks for your work!

**Comment by Innei** [comment]
Is there any progress now?

**Comment by evolify** [comment]
Nothing update, I have go back to neovim.

**Comment by macintacos** [comment]
Perhaps this just needs to be reassigned? @berknam seems to not be very active anymore. Considering that we've identified the commit that caused this issue in [this comment](https://github.com/VSCodeVim/Vim/issues/5884#issuecomment-766413548), perhaps the fix is straightforward? As far as I can tell this is still an issue in the latest version of the extension.

@J-Fields any chance this could get re-prioritized?

**Comment by djgray780** [comment]
It might be worth a try, but I have the following Emmet settings right now and I can use the abbreviations without getting shoved into visual mode.

`    //* Emmet Settings *//
    "emmet.triggerExpansionOnTab": true,
    "emmet.showExpandedAbbreviation": "always",
    "emmet.includeLanguages": {
        "javascript": "javascriptreact"
    }`

**Comment by rehack** [comment]
您好！您的邮件已经收到，我会及时回复您，感谢您对睿客网的支持。需了解更多高质量的web开发技术请移步：http://www.rehack.cn/

**Comment by chang2000** [comment]
> It works. Thanks a lot

**Comment by rtalexk** [comment]
I had faced the same issue.
After doing some research to configure VSCode for Rails, I've fixed this issue by doing:
* Install the [Ruby extension](https://marketplace.visualstudio.com/items?itemName=rebornix.Ruby).
* Add `"erb": "html"` to the `emmet.includeLanguages` array setting.

Here's my full settings file in case you wanna replicate it:
```JSON
{
    "editor.fontSize": 16,
    "workbench.colorTheme": "Cobalt2",
    "editor.bracketPairColorization.enabled": true,
    "editor.bracketPairColorization.independentColorPoolPerBracketType": true,
    "editor.guides.bracketPairs": true,
    "javascript.format.insertSpaceAfterOpeningAndBeforeClosingNonemptyBrackets": true,
    "typescript.format.insertSpaceAfterOpeningAndBeforeClosingNonemptyBrackets": true,
    "editor.tabSize": 2,
    "editor.fontFamily": "'Fira Code', Menlo, Monaco, 'Courier New', monospace",
    "editor.fontLigatures": true,
    "editor.copyWithSyntaxHighlighting": false,
    "emmet.includeLanguages": {
        "erb": "html",
        "html": "html",
        "html.erb": "html",
        "ruby": "html",
        "html.inky": "html"
    },
    "emmet.triggerExpansionOnTab": true,
    "emmet.showExpandedAbbreviation": "always",
    "workbench.iconTheme": "vscode-great-icons",
    "auto-close-tag.fullMode": true,
    "editor.minimap.enabled": false,
    "files.associations": {
        "*.erb": "erb",
        "Gemfile": "ruby"
    }
}
```

And the extensions I have installed:
<img width="401" alt="image" src="https://user-images.githubusercontent.com/16341644/164931676-99659473-bf30-4b5a-9bd5-85871fdbd863.png">

\* Most of them are not related to the issue, but may be worth mention it.

**Comment by pingshunhuangalex** [comment]
Can confirm this is still an issue in:
```
Version: 1.68.0-insider (user setup)
Commit: c23f0305dbf82b2319b198f4dbf3c5d5bc522f15
Date: 2022-05-16T05:17:17.823Z
Electron: 17.4.3
Chromium: 98.0.4758.141
Node.js: 16.13.0
V8: 9.8.177.13-electron.0
OS: Windows_NT x64 10.0.19044
```

**Comment by mrwest808** [comment]
I'm experimenting with this issue right now (locally in my settings), and I've noticed that this abomination of a custom keybinding actually seems to produce the desired effect.

I simply tried recreating the steps I usually do manually to counteract this issue (`<Esc>` to exit visual mode followed by `c i t` to enter input mode inside of the newly created tag).

```json
"vim.insertModeKeyBindingsNonRecursive": [
  {
    "before": ["<C-e>"],
    "after": [],
    "commands": [
      {
        "command": "editor.emmet.action.expandAbbreviation",
        "args": []
      },
      {
        "command": "extension.vim_escape",
        "args": []
      },
      {
        "command": "extension.vim_insert",
        "args": []
      },
      {
        "command": "extension.vim_right",
        "args": []
      },
      {
        "command": "extension.vim_right",
        "args": []
      },
      {
        "command": "extension.vim_right",
        "args": []
      }
    ]
  }
]
```

**Comment by rehack** [comment]
您好！您的邮件已经收到，我会及时回复您，感谢您对睿客网的支持。需了解更多高质量的web开发技术请移步：http://www.rehack.cn/

**Comment by tiborsimon** [comment]
This issue is still relevant. I modified a bit @mrwest808 solution to be able to handle more scenarios with a simple repeated keystroke:

```json
{
  "vim.insertModeKeyBindingsNonRecursive": [
    {
      "before": ["<C-m>"],
      "after": [],
      "commands": [
        {
          "command": "editor.emmet.action.expandAbbreviation",
          "args": []
        },
        {
          "command": "extension.vim_escape",
          "args": []
        }
      ]
    },
    ...
  ],
  "vim.normalModeKeyBindingsNonRecursive": [
    {
      "before": ["<C-m>"],
      "after": ["f", "/", "h", "i"]
    },
    ...
  ],
  ...
}
```

It will map the <kbd>C-m</kbd> key combination (_<kbd>C-m</kbd> is more ergonomic for me_) in both insert mode and in normal mode. The mapping in insert mode would trigger `emmet` and exits to vim normal mode, then the next mapping in normal mode would navigate to the most-inner tag that is generated. You just have to press <kbd>C-m</kbd> twice, and you are golden 👯 

I cannot found a solution to do it with one mapping as `vscodevim` doesn't support the `:normal` command yet, and I don't want to bother with integrating the external `nvim` to `VSCode` yet.

With this double mapping you can have more complex `emmet` expansions:

```
# The @ character is the cursor position

p.subtitle@
[C-m][C-m]
<p class="subtitle">@</p>

a>span.icon.github@
[C-m][C-m]
<a href=""><span class="icon github">@</span></a>
```

**Comment by evolify** [comment]
Any update?
